File: euler/done/004.rb
#!/usr/bin/ruby

=begin rdoc

http://projecteuler.net/index.php?section=problems&id=4

A palindromic number reads the same both ways. The largest palindrome
made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit
numbers.

=end

def is_palindromic(n)
    str = '%d' % n

    return str == str.reverse
end

max = 0
100.upto(999) { |x|
    100.upto(999) { |y|
        n = x*y
        if is_palindromic(n)
            #print "#{x} x #{y} = #{n}\n"
            max = n if n > max
        end
    }
}
print max, "\n"

File: euler/done/004.ex
defmodule Euler004 do
  # https://projecteuler.net/problem=4
  # A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
  # Find the largest palindrome made from the product of two 3-digit numbers.

  def foo() do
    palindromes = for x <- 1..999, y <- 1..999, is_palindromic?( x * y ), do: x*y
    palindromes |> Enum.max
  end


  def is_palindromic?( n ) do
    digits = digits( n )
    digits == Enum.reverse( digits )
  end

  def digits( n ) do
    _digits( [n] )
  end

  defp _digits( [ n | digits ] ) when n > 0 and n < 10 do
    [ n | digits ]
  end

  defp _digits( [ n | digits ] ) when n >= 10 do
    _digits( [ div(n,10), rem(n,10) | digits ] )
  end

end
